modules/src/core/ics02_client/client_def.rs

```diff
@@ -16,63 +13,73 @@ use crate::core::ics23_commitment::commitment::{
 };
 use crate::core::ics24_host::identifier::{ChannelId, ClientId, ConnectionId, PortId};
 use crate::core::ics26_routing::context::ReaderContext;
-use crate::downcast;
 use crate::prelude::*;
 use crate::Height;
 use core::fmt::Debug;
 
-#[cfg(any(test, feature = "mocks"))]
-use crate::mock::client_def::MockClient;
-
 #[derive(PartialEq, Eq, Clone, Debug)]
-pub enum ConsensusUpdateResult {
-    Single(AnyConsensusState),
-    Batch(Vec<(Height, AnyConsensusState)>),
+pub enum ConsensusUpdateResult<C: ClientKeeper> {
+    Single(C::AnyConsensusState),
+    Batch(Vec<(Height, C::AnyConsensusState)>),
+}
+
+impl<C: ClientKeeper> ConsensusUpdateResult<C> {
+    pub fn map_state<F, D: ClientKeeper>(self, f: F) -> ConsensusUpdateResult<D>
+    where
+        F: Fn(C::AnyConsensusState) -> D::AnyConsensusState,
+    {
+        match self {
+            ConsensusUpdateResult::Single(cs) => ConsensusUpdateResult::Single(f(cs)),
+            ConsensusUpdateResult::Batch(cs) => {
+                ConsensusUpdateResult::Batch(cs.into_iter().map(|(h, s)| (h, f(s))).collect())
+            }
+        }
+    }
 }
 
 pub trait ClientDef: Clone {
     type Header: Header;
-    type ClientState: ClientState;
-    type ConsensusState: ConsensusState;
+    type ClientState: ClientState + Eq;
+    type ConsensusState: ConsensusState + Eq;
 
-    fn verify_header(
+    fn verify_header<Ctx: ReaderContext>(
         &self,
-        ctx: &dyn ReaderContext,
+        ctx: &Ctx,
         client_id: ClientId,
         client_state: Self::ClientState,
         header: Self::Header,
     ) -> Result<(), Error>;
 
-    fn update_state(
+    fn update_state<Ctx: ReaderContext>(
         &self,
-        ctx: &dyn ReaderContext,
+        ctx: &Ctx,
         client_id: ClientId,
         client_state: Self::ClientState,
         header: Self::Header,
-    ) -> Result<(Self::ClientState, ConsensusUpdateResult), Error>;
+    ) -> Result<(Self::ClientState, ConsensusUpdateResult<Ctx>), Error>;
 
     fn update_state_on_misbehaviour(
         &self,
         client_state: Self::ClientState,
         header: Self::Header,
     ) -> Result<Self::ClientState, Error>;
 
-    fn check_for_misbehaviour(
+    fn check_for_misbehaviour<Ctx: ReaderContext>(
```

Comment:
This function is meant to be used when handling misbehaviour messages, we should refactor it to accept take header_1 && header_2, not just one header.

Reply:
don't think this necessary, because of how 02-client uses it:

https://github.com/ComposableFi/ibc-rs/blob/04bdb0ddfea674a5f77ca50ba97d468a481d4d75/modules/src/core/ics02_client/handler/update_client.rs#L85-L107

Reply:
Yeah, it's not.

Reply:
so it means it shouldn't be `verrify_header` but instead, it should be `verify_client_message`

https://github.com/cosmos/ibc-go/blob/80ea89a356da6606b351f0e00fc82786a5c13fad/modules/light-clients/07-tendermint/update.go#L21-L33

Reply:
@Wizdave97 can you make the necessary refactor on this PR?